### ISSUE TITLE
Exception during XNB content loading : mismatch between "mscorlib" and "System.Private.CoreLib"

### ISSUE THREAD

**Issue by Paphos** [MonoGame/MonoGame#7872]
Hello!

I have migrated my projects to MonoGame 3.8.1.303 and everything is building correctly.

Unfortunately, at runtime, when I am trying to load XNB content from the ContentManager...

```csharp
ArmatureData armature = _monogameContentManager.Load<ArmatureData>(myArmatureName);
```

... I get this error :

```
Microsoft.Xna.Framework.Content.ContentLoadException: 'Could not find ContentTypeReader Type.
Please ensure the name of the Assembly that contains the Type matches the assembly in the full type name:
Microsoft.Xna.Framework.Content.ArrayReader'1[[System.Numerics.Matrix4x4, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
(Microsoft.Xna.Framework.Content.ArrayReader'1[[System.Numerics.Matrix4x4, mscorlib]])'
```

I think the problem comes from a mismatch between `System.Private.CoreLib` and `mscorlib`. My whole solution is in .NET 6, and so are the MonoGame packages, so there shouldn't be any references to `mscorlib`, am I correct?

# Context

I am trying to load this type :

```csharp
public class ArmatureData
{
    public Matrix4x4[] BindPose { get; set; }
    public Matrix4x4[] InverseBindPose { get; set; }
    public int[] SkeletonHierarchy { get; set; }
}
```

As you can see, I am using `System.Numerics.Matrix4x4` and not the `Microsoft.Xna.Framework.Matrix` but it never caused any problems until today.

My solution has 3 projects (all in `net6.0-windows`) :
- `MyGame.Shared`, a class library in `net6.0-windows` that does not have any package or reference, and containing shared models including my `ArmatureData` class.
- `MyGame.MgcbPipelineExtension`, a class library in `net6.0-windows` used dynamically by the MGCB Builder Tool, and that references `MyGame.Shared` and these packages :
```xml
    <PackageReference Include="MonoGame.Framework.Content.Pipeline" Version="3.8.1.303" />
    <PackageReference Include="MonoGame.Framework.WindowsDX" Version="3.8.1.303" />
```
- `MyGame.App`, the main executable in `net6.0-windows` that references `MyGame.Shared` and these packages :
```xml
    <PackageReference Include="MonoGame.Framework.WindowsDX" Version="3.8.1.303" />
    <PackageReference Include="MonoGame.Content.Builder.Task" Version="3.8.1.303" />
```

# Possible solution?

I think the mismatch between `System.Private.CoreLib` and `mscorlib` comes from here :

https://github.com/MonoGame/MonoGame/blob/574fd7f870a90d4c85ea6776acda359b08874d2a/MonoGame.Framework/Content/ContentTypeReaderManager.cs#L229-L232

The detection of whether we are in .NET Core doesn't seem to work, but I am not really sure : 
https://github.com/MonoGame/MonoGame/blob/574fd7f870a90d4c85ea6776acda359b08874d2a/MonoGame.Framework/Content/ContentTypeReaderManager.cs#L24

When I am loading XNB content at runtime (in .NET 6), it looks like `_isRunningOnNetCore` is set to `false` for a strange reason, and this line wrongfully replace `System.Private.CoreLib` by `mscorlib` in the fully qualified name of the `Matrix4x4` type.

In any case, since we are now in .NET 6 only, could we just remove these lines ?
Or maybe just keep this one for compatibility purpose ?

https://github.com/MonoGame/MonoGame/blob/574fd7f870a90d4c85ea6776acda359b08874d2a/MonoGame.Framework/Content/ContentTypeReaderManager.cs#L230

I am not really familiar with these parts so I am not sure what it could break, but it seems to be the solution to my problem.

What do you think?

**Comment by mrhelmut** [comment]
Duplicate of #7868 and it isn't quite an issue of MonoGame. You are likely using trimming or trying to AOT compile, which breaks reflection. Please [disable trimming or make sure that you configured your project to not trim MonoGame](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options).

I'll be closing this in favor of  #7868 while we work on adding a note about trimming to the documentation, even though it isn't specific to MonoGame.

**Comment by Paphos** [comment]
Thank you for your reply!
I am not using any compilation option (neither trimming nor AOT) for these projects (expect `CopyLocalLockFileAssemblies` for my "MgcbPipelineExtension" project, but I have the same error without it), so I don't think this is the reason.
But I will try to find a workaround.
Thank you again! 😊

**Comment by mrhelmut** [comment]
> Microsoft.Xna.Framework.Content.ContentLoadException: 'Could not find ContentTypeReader Type.

This typically is a trimming error, like 99% chances that it is one. I would suggest looking into that direction, otherwise I fear that you'd be chasing rainbows. Let me know if you find anything relevant.

**Comment by harry-cpp** [comment]
Oh I remember this error: https://github.com/MonoGame/MonoGame/pull/7199/files?w=1

TLDR. is mgcb produces different type strings between running it on .NET Framework and .NET, so the type written might not match the targeted platform, those lines were added both for backwards compatibility of content, as well as allowing you to load stuff no matter how you compile mgcb.

Interesting that the `Type.GetType("System.Private.CoreLib")` is returning false, maybe the assembly was not loaded before this line was reached therefore the type is not available 🤔 

Anyhow reopening for now so I remember to take a look at it tomorrow!

**Comment by vpenades** [comment]
It could be possible to decorate the Load<> method with this attribute:

```c#
[System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("Disable trimming")]
public static T DeserializeWithReflection<T>(Object value)
{
// some reflection code
}
```

`RequiresUnreferencedCode` is a newly introduced attribute to deal with these case scenarios.  If a developer uses reflection, AND trimming on types that are not readily available, it will trigger warnings by the analyzers.

**Comment by mrhelmut** [comment]
FYI ```_isRunningOnNetCore``` will be ```false``` on consoles (with BRUTE at least, likely ```true``` with NativeAOT) while content is likely to be built with newer runtimes, so we kind of need that check (I was about to suggest to drop backward compatibility, but I don't think that we can).

(I concede that I didn't look closely enough at this issue at first, it is indeed not a trimming issue, but it looked like it at first glance.)

**Comment by Paphos** [comment]
No problem!

In the end, I bypassed the problem by building/loading the few problematic assets directly from my code, and not by using the MGCB or ContentManager. 😅
Not really a solution, of course, but for my specific needs, it was acceptable. And from there, the migration of my code to 3.8.1 went smoothly! 🥳

Thank you all for your replies, and for the great work in this release! 🌸

**Comment by eosxff** [comment]
I do wonder if something like `private static readonly bool _isRunningOnNetCore = typeof(object).Assembly.GetName().Name == "System.Private.CoreLib"` would work as a replacement since the original is failing for some odd reason.

**Comment by eosxff** [comment]
I can also confirm Type.GetType("System.Private.CoreLib") returns null on net6.0-windows, I'll make a PR

**Comment by ossyahya60** [comment]
Hi, I faced the same issue and I managed to solve it by basically adding the mscorlib.dll to the output directory (bin folder)
